docs/design/sql/20-match-jet-job-to-sql.md

```diff
@@ -0,0 +1,189 @@
+additional field in `SqlResult` (or `SqlClientResult`) class, which will contain `jobId` of the corresponding Jet job.
+
+This seemingly minor alteration, when examined closely, carries significant implications:
+we must now match Jet jobs with SQL queries during query execution, ideally even before the Jet job start its execution.
+The challenge with this approach is that presently (as of HZ 5.3), for small batch queries, as well as for all DML
+queries, the `SqlResult` is not available until after the job has finished. To address this predicament, we need
+to rework our query lifecycle to operate asynchronously: we'll return an empty `SqlResult` (or empty analogue
+of `SqlResult`) to the client and prompt the client to `fetch` the results. Worth noticing, it also solves a problem
+of query thread blocking during DML query execution.
+Overall, this way causes a substantial shift for a metrics-related task as a short term goal, necessitating considerable
+effort and time for implementation. But for a long-running perspective it seems reasonable effort.
+
+#### Solution design
+
+##### Client protocol
+
+To enable asynchronous query handling, we propose adding a field to the `SqlExecuteCodec.RequestParameters` to mark 
+requests that should be handled asynchronously. In order to support [Flexible Async Mode](#flexible-async-mode)
+it will be an integer field with the following values:
+- `-1` - synchronous mode (default and for backwards compatibility)
+- `0` - asynchronous mode
+- `> 0` - timeout for flexible async mode before returning async response
+
+Initially only the first 2 values will be supported. Flexible async mode may be implemented in the future if the performance
+tests confirm that latency for short DML queries might be improved that way.
+
+##### Java APIs
+
+In the `SqlService`, we will introduce a new method called `executeAsync`. This method will set the asynchronous flag
+when executing the request, signaling that the query should be handled asynchronously. After executing a query using
+`executeAsync`, the client should check the first arrived result to determine if it produces rows. If rows are produced,
+it will continue with normal execution. If not (user wants to know when the query finishes), then `job.join()` must 
+be explicitly called. For better user experience, job joining may be encapsulated in new `AsyncSqlResult.join()` method.
+
+Note that async execution will not be a default mode for the time being because without flexible async mode
+it may increase latency for short DML queries due to additional roundtrip to the member.
+
+```java
+public interface SqlService {
+    // ...
+
+    /**
+     * Execute the statement in asynchronous way if possible.
+     * @param statement
+     * @return
+     */
+    // also additional convenience variants similar to `execute`
+    // supported only on client-side (SqlClientService).
```

Comment:
@Fly-Style does it make sense to support async in member API? There might be some exotic use cases (invoking SQL from member-side code like Map listeners, Executors) but is it worth the complication? Are there any other use cases for that?